### PR TITLE
Add Evaluation of Pool Query Strategies

### PR DESCRIPTION
**Pull request by Cheng-JY**
1. Add a Evaluation notebook of 6 poolbased query Strategies:

- Random Sampling
- Uncertainty Sampling
- Discriminative AL
- Core Set
- TypiClust
- Badge

2. This Evaluation based on pre-processing dataset ```Flowers102``` with ```Dinov2```

3. The Evaluation results is illustrated with a learning curves

Side Consideration:
I have changed the query_batch_size from 32 to 16, so that the learning curve looks better and it does not immediately reach its upper bound.

### CODE REVIEW

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9c0819f4-b45f-4285-aa1d-84469764b6a5",
+   "metadata": {},
+   "source": [
+    "**Step 1: Transformation**\n",
+    "\n",
+    "Apply necessary transformations to the datasets, including resizing images to a standardised format. This ensures consistency of input dimensions to the DINOv2 model."
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:145]:
```suggestion
    "Apply necessary transformations to the datasets, including resizing images to a standardized format. This ensures consistency of input dimensions with the DINOv2 model."
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+   "outputs": [],
+   "source": [
+    "transforms = transforms.Compose(\n",
+    "        [transforms.Resize(256),\n",
+    "         transforms.CenterCrop(224),\n",
+    "         transforms.ToTensor(),\n",
+    "         transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225))]\n",
+    "    )\n",
+    "\n",
+    "batch_size = 4"
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:168]:
```suggestion
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "8c0165be-d3ce-4eb2-9309-45ba6c08c8d3",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def load_and_process_dataset(dataset_name, root_dir, is_train):\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:262]:
```suggestion
    "def load_and_process_dataset(root_dir, is_train, batch_size=4):\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "dinov2_vitb14 = torch.hub.load(\"facebookresearch/dinov2\", \"dinov2_vitb14\")\n",
+    "\n",
+    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
+    "\n",
+    "dinov2_vitb14.to(device)"
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:238]:
```suggestion
    "hub.set_dir('cache/')\n",
    "dinov2_vitb14 = torch.hub.load(\"facebookresearch/dinov2\", \"dinov2_vitb14\")\n",
    "\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "\n",
    "dinov2_vitb14.to(device)"
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "7668a073b61085ea",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "**Step 3: Load Datasets and Generate Embeddings**\n",
+    "\n",
+    "Firstly, we begin by loading the CIFAR-10, CIFAR-100, and Flowers-102 datasets.\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:250]:
```suggestion
    "Firstly, we begin by loading the Flowers-102 dataset.\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+    "\n",
+    "After that, we employ the pre-trained DINOv2 model to generate embeddings for each image in the datasets and save them in a npy file."
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:252]:
```suggestion
    "After that, we employ the pre-trained DINOv2 model to generate embeddings for each image in the dataset and save them in a npy file."
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+   "outputs": [],
+   "source": [
+    "def load_and_process_dataset(dataset_name, root_dir, is_train):\n",
+    "    \"\"\"\n",
+    "    Load and process a given dataset for training or validation.\n",
+    "\n",
+    "    Parameters:\n",
+    "    - dataset_name (str): Name of the dataset (e.g., 'CIFAR10', 'CIFAR100').\n",
+    "    - root_dir (str): Root directory where the dataset will be stored.\n",
+    "    - is_train (bool): Boolean indicating whether the dataset is for training (True) or validation (False).\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:269]:
```suggestion
    "    Parameters:\n",
    "    - root_dir (str): Root directory where the dataset will be stored.\n",
    "    - is_train (bool): Boolean indicating whether the dataset is for training (True) or validation (False).\n",
    "    - batch_size (int): The batch_size used for the DataLoader.\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+    "    - y_true (numpy.ndarray): Concatenated true labels of the dataset.\n",
+    "    \"\"\"\n",
+    "    \n",
+    "    # Load the dataset\n",
+    "    if is_train:\n",
+    "        split = 'train'\n",
+    "    else:\n",
+    "        split = 'val'\n",
+    "    \n",
+    "    dataset = datasets.__dict__[dataset_name](root=root_dir, split=split, download=True, transform=transforms)\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:282]:
```suggestion
    "    dataset = datasets.Flowers102(root=root_dir, split=split, download=True, transform=transforms)\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+    "        \n",
+    "\n",
+    "    # Create a DataLoader\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:285]:
```suggestion
    "    # Create DataLoader\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+    "    - dataset_name (str): Name of the dataset (e.g., 'CIFAR10', 'CIFAR100').\n",
+    "    - root_dir (str): Root directory where the dataset will be stored.\n",
+    "    - is_train (bool): Boolean indicating whether the dataset is for training (True) or validation (False).\n",
+    "\n",
+    "    Returns:\n",
+    "    - X (numpy.ndarray): Concatenated embeddings of the dataset.\n",
+    "    - y_true (numpy.ndarray): Concatenated true labels of the dataset.\n",
+    "    \"\"\"\n",
+    "    \n",
+    "    # Load the dataset\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:276]:
```suggestion
    "    # Load dataset\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+    "import torch\n",
+    "import torchvision.datasets as datasets\n",
+    "import torchvision.transforms as transforms\n",
+    "from tqdm import tqdm\n",
+    "\n",
+    "import warnings\n",
+    "mlp.rcParams[\"figure.facecolor\"] = \"white\"\n",
+    "warnings.filterwarnings(\"ignore\")"
+   ]
+  },
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:115]: Add a cell that defines a variable named 'data_dir' that points to "./data"

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+   ],
+   "source": [
+    "# Flowers-102\n",
+    "flowers102_X_train, flowers102_y_train_true = load_and_process_dataset(\"Flowers102\", \"./data\", True)\n",
+    "flowers102_X_test, flowers102_y_test_true = load_and_process_dataset(\"Flowers102\", \"./data\", False)\n",
+    "\n",
+    "np.save('./embedding_data/flowers102_dinov2B_X_train.npy', flowers102_X_train)\n",
+    "np.save('./embedding_data/flowers102_dinov2B_y_train.npy', flowers102_y_train_true)\n",
+    "np.save('./embedding_data/flowers102_dinov2B_X_test.npy', flowers102_X_test)\n",
+    "np.save('./embedding_data/flowers102_dinov2B_y_test.npy', flowers102_y_test_true)"
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:337]: export to data_dir as well

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+   "execution_count": 7,
+   "id": "e29fc457-2fb1-4824-8cd5-b637e44795a8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Flowers-102\n",
+    "X_train = np.load('./embedding_data/flowers102_dinov2B_X_train.npy')\n",
+    "y_train_true = np.load('./embedding_data/flowers102_dinov2B_y_train.npy')\n",
+    "X_test = np.load('./embedding_data/flowers102_dinov2B_X_test.npy')\n",
+    "y_test_true = np.load('./embedding_data/flowers102_dinov2B_y_test.npy')\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:361]: Use data_dir

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+   "source": [
+    "## Classification Models and Query Strategies"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "91a3743e-cb9e-49a9-a239-30c9f54839fa",
+   "metadata": {},
+   "source": [
+    "The embeddings we have computed can be used as an input in a classification model. For this guide, we will be using LogisticRegression from ```sklearn``` [5]. Moreover, we handle the creation of query strategies using factory functions to simplify the seperation.\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:439]:
```suggestion
    "The embeddings we have computed can be used as an input to a classification model. For this guide, we use `LogisticRegression` from `scikit-klearn`. Moreover, we handle the creation of query strategies using factory functions to simplify the seperation of classifiers and query strategies across repetitions and folds.\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+    "\n",
+    "[5] ‘sklearn.linear_model.LogisticRegression’, scikit-learn. Accessed: Jan. 24, 2024. [Online]. Available: https://scikit-learn/stable/modules/generated/sklearn.linear_model.LogisticRegression.html"
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:441]:
```suggestion
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 16]
```diff
@@ -0,0 +1,626 @@
...
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-01-17T16:23:51.265616Z",
+     "start_time": "2024-01-17T16:23:51.249552Z"
+    },
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "clf = SklearnClassifier(LogisticRegression(), classes=np.arange(dataset_classes), random_state=gen_seed(master_random_state))\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:457]: Add a line break to shorten this line

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "9023c37048ec8f54",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Experiment Parameters\n",
+    "\n",
+    "For this experiment, we need to define how the strategies should be compared against one another. Here the number of repetitions (```n_reps```), the numer of queries (```n_cycles```), the size of each query (```query_batch_size```) need to be difined. "
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:481]:
```suggestion
    "For this experiment, we need to define how the strategies should be compared against one another. Here the number of repetitions (```n_reps```), the number of cycles(```n_cycles```), and the size of each query (```query_batch_size```) need to be defined. "
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "e103670190e73a3f",
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "n_reps = 1\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:493]:
```suggestion
    "n_reps = 3\n",
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "66b8fff1b12211ed",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Experiment Loop\n",
+    "\n",
+    "The actual experiment loops all over the query strategies. The average accuracy over the test set is recorded for each cycle and stored in the ```results``` dictionary."
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:508]:
```suggestion
    "The actual experiment loops over all query strategies. The accuracy for the test set is stored for each cycle and repetition in the `results` dictionary."
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 20]
```diff
@@ -0,0 +1,626 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "998e52cb1a2107b",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Resulting Plotting\n",
+    "\n",
+    "We use learning curves to compare strategies. We visualise the average accuracy, averaged over all repetitions and folds, relative to the number of queries. In addition, the legend provides insight into the area under the learning curve, which indicates the average accuracy over all cycles."
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:560]:
```suggestion
    "We use learning curves to compare strategies. We visualize the average accuracy over all repetitions. In addition, the legend provides insight into the area under the learning curve, which indicates the average accuracy over all cycles."
```

File: tutorials/05_pool_evaluation_with_self_supervised.ipynb [hunk 21]
```diff
@@ -0,0 +1,626 @@
...
+    }
+   ],
+   "source": [
+    "for qs_name in query_strategy_names:\n",
+    "    key = qs_name\n",
+    "    result = results[key]\n",
+    "    reshaped_result = result.reshape((-1, n_cycles))\n",
+    "    errorbar_mean = np.mean(reshaped_result, axis=0)\n",
+    "    plt.errorbar(np.arange(n_cycles), errorbar_mean, label=f\"({np.mean(errorbar_mean):.4f}) {qs_name}\", alpha=0.5)\n",
+    "plt.title(\"LogisticRegression\")\n",
```
Comment (tpham93) [inline tutorials/05_pool_evaluation_with_self_supervised.ipynb:589]:
```suggestion
    "plt.title(f\"LogisticRegression with query batch size {query_batch_size}\")\n",
```

**Review by tpham93 — CHANGES_REQUESTED** [review]

File: tutorials/05_pool_al_with_self_supervised_learning.ipynb [hunk 22]
```diff
@@ -0,0 +1,676 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "9023c37048ec8f54",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Experiment Parameters\n",
+    "\n",
+    "For this experiment, we need to define how the strategies should be compared against one another. Here the number of repetitions (```n_reps```), the numer of queries (```n_cycles```), the size of each query (```query_batch_size```) need to be difined. "
```
Comment (tpham93) [inline tutorials/05_pool_al_with_self_supervised_learning.ipynb:507]:
```suggestion
    "For this experiment, we need to define how the strategies should be compared against one another. Here the number of repetitions (```n_reps```), the number of cycles(```n_cycles```), and the size of each query (```query_batch_size```) need to be defined. "
```

File: tutorials/05_pool_al_with_self_supervised_learning.ipynb [hunk 23]
```diff
@@ -0,0 +1,676 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "66b8fff1b12211ed",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Experiment Loop\n",
+    "\n",
+    "The actual experiment loops all query strategies. The accuracy for the test set is stored for each cycle and repetition in the `results` dictionary."
```
Comment (tpham93) [inline tutorials/05_pool_al_with_self_supervised_learning.ipynb:538]:
```suggestion
    "The actual experiment loops over all query strategies. The accuracy for the test set is stored for each cycle and repetition in the `results` dictionary."
```

File: tutorials/05_pool_al_with_self_supervised_learning.ipynb [hunk 24]
```diff
@@ -0,0 +1,676 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "998e52cb1a2107b",
+   "metadata": {
+    "collapsed": false
+   },
+   "source": [
+    "## Resulting Plotting\n",
+    "\n",
+    "We use learning curves to compare strategies. We visualize the average accuracyover all repetitions. In addition, the legend provides insight into the area under the learning curve, which indicates the average accuracy over all cycles."
```
Comment (tpham93) [inline tutorials/05_pool_al_with_self_supervised_learning.ipynb:607]:
```suggestion
    "We use learning curves to compare strategies. We visualize the average accuracy over all repetitions. In addition, the legend provides insight into the area under the learning curve, which indicates the average accuracy over all cycles."
```

**Review by mherde — APPROVED** [review]

**Pull request merged**